File: Android/app/src/main/java/com/nfsindustries/cryptocurrencymonitor/deserializer/CoinmarketcapDeserializer.java
package com.nfsindustries.cryptocurrencymonitor.deserializer;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.nfsindustries.cryptocurrencymonitor.model.CryptoCurrencyModel;

import java.lang.reflect.Type;

/**
 * Created by maxmo on 26/5/2017.
 */

public class CoinmarketcapDeserializer implements JsonDeserializer<CryptoCurrencyModel> {

    @Override
    public CryptoCurrencyModel deserialize(final JsonElement json, final Type typeOfT,
                                    final JsonDeserializationContext context)
            throws JsonParseException {

        final JsonArray rootJsonArray =  json.getAsJsonArray();
        final JsonElement usdJson = rootJsonArray.get(0);

        final Gson gson = new Gson();
        final CryptoCurrencyModel cryptoCurrencyModel = gson.fromJson(usdJson, CryptoCurrencyModel.class);

        return cryptoCurrencyModel;
    }
}

File: Android-Kotlin/app/src/main/java/com/nfsindustries/cryptocurrencymonitor/deserializer/CoinmarketcapDeserializer.kt
package com.nfsindustries.cryptocurrencymonitor.deserializer

import com.google.gson.Gson
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.nfsindustries.cryptocurrencymonitor.model.CryptoCurrencyModel

import java.lang.reflect.Type

/**
 * Deserializer for CoinMarketCap API, used to convert data from JSON
 * It uses Google's GSON parsing library
 * For now, it only converts USD [United States Dollar]
 * TODO add support for other countries' currencies
 */

class CoinmarketcapDeserializer : JsonDeserializer<CryptoCurrencyModel> {

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type,
                             context: JsonDeserializationContext): CryptoCurrencyModel {

        val rootJsonArray = json.asJsonArray
        val usdJson = rootJsonArray.get(0)

        val gson = Gson()

        return gson.fromJson(usdJson, CryptoCurrencyModel::class.java)
    }
}